llext: add STT_OBJECT relocation

laurenmurphyx64 opened the pull request:
Adds STT_OBJECT relocation.

Fixes #67701.

You can test with the attached patch (it's a patch, just renamed) [0001-STT_OBJECT_test.patch.log](https://github.com/zephyrproject-rtos/zephyr/files/14397787/0001-STT_OBJECT_test.patch.log) and the `tests/subsys/llext/hello_world` test.

laurenmurphyx64 committed:
llext: add STT_OBJECT relocation

Adds STT_OBJECT relocation.

Fixes #67701.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

teburd commented on subsys/llext/llext.c:
```diff
@@ -776,6 +777,7 @@ static int llext_link(struct llext_loader *ldr, struct llext *ext, bool do_local
 				LOG_INF("found section symbol %s addr 0x%lx", name, link_addr);
 			} else {
 				/* Nothing to relocate here */
+				LOG_INF("not relocated");
```
I'd say this should likely be a LOG_DBG

laurenmurphyx64 replied:
> The base ztest suite should be the same, and run through all the available test extensions like the simplest hello world printk, this sem give/take with a global, and so on.

@teburd I've put one up, have a look and see if you like it / let me know what other test cases you'd like to see.

teburd reviewed (commented):
Change looks good, I think it could us with a test case

copying the hello_world extension and updating it to work a little different and do what your patch does with a global (non-static global) variable should do the trick.

The base ztest suite should be the same, and run through all the available test extensions like the simplest hello world printk, this sem give/take with a global, and so on.

laurenmurphyx64 commented:
> The base ztest suite should be the same, and run through all the available test extensions like the simplest hello world printk, this sem give/take with a global, and so on.

@teburd I've put one up, have a look and see if you like it / let me know what other test cases you'd like to see.

pillo79 reviewed (commented):
Change looks obviously good and thanks for the test code!
 
Personally, I'd like `hello world` to be the simplest possible example. I would therefore create a separate `thread_test` project integrating `CONFIG_USERSPACE` and your new thread test code, and removing the duplicate code from the original `hello_world`. @teburd what's your take?

teburd commented:
> Change looks obviously good and thanks for the test code!
> 
> Personally, I'd like `hello world` to be the simplest possible example. I would therefore create a separate `thread_test` project integrating `CONFIG_USERSPACE` and your new thread test code, and removing the duplicate code from the original `hello_world`. @teburd what's your take?

This can be done with testcase.yml, consolidating the tests into one test project is helpful honestly. The more of these tests we have that are seperate they tend to add to actual test time primarily from building and flashing rather than actually running test code.


> Change looks obviously good and thanks for the test code!
> 
> Personally, I'd like `hello world` to be the simplest possible example. I would therefore create a separate `thread_test` project integrating `CONFIG_USERSPACE` and your new thread test code, and removing the duplicate code from the original `hello_world`. @teburd what's your take?

The test was named simple to exercise simple extensions and their code paths, this is still a pretty simple extension and I think a worthwhile addition here.

The trouble with making too many separate test applications is the build and flash time grows faster than the time it takes to run the tests by a pretty large margin sometimes, especially for something like this.

If separately compiled tests do become necessary we can use testcase.yaml to decide which extension to test while utilizing all the test code structure still. Personally I don't think it'll be necessary though, the extensions themselves are small.

pillo79 commented:
Thanks for the explanations. 

---

Looked at CI failures - they are due to the `-mcpu=cortex33+nodsp` flag that llext forces when compiling ARM extensions. That has been done in the early commits to force the llext compiler to use a fixed CPU instruction set (I _think_ to disable complex jump opcodes from being emitted?), but it's obviously a hack that should be addressed.

cfriedt commented:
@laurenmurphyx64 - looks good.

Just one question - do we already have exported syscalls pr exported symbols for `k_sem_init()`, `k_thread_join()`, etc?

There are likely quite a number of thread-related exports that might be necessary in the long run, but if we're exporting `k_thread_create()`, we should probably simultaneously export `k_thread_join()` at least.

Are there any other thread APIs that would be safe to complement `k_thread_create()`? E.g. `k_thread_abort()`?

laurenmurphyx64 commented:
> There are likely quite a number of thread-related exports that might be necessary in the long run, but if we're exporting `k_thread_create()`, we should probably simultaneously export `k_thread_join()` at least.

@cfriedt I'm working on adding k_thread_join, but trying to use it causes faults ("unaligned memory access"). Hopefully k_thread_abort won't do the same thing when I try it

teburd commented:
Please rebase, should fix CI issues

laurenmurphyx64 committed:
llext: add object test case

Adds a case to the simple llext test testing global
non-static object relocation.

Signed-off-by: Lauren Murphy <lauren.murphy@intel.com>

laurenmurphyx64 commented:
Revised to simplify example to just object relocation to keep it in scope of the PR, threads to come in another PR for a bug report. Should be ready to merge assuming CI succeeds...

teburd commented on tests/subsys/llext/simple/src/object_ext.c:
```diff
@@ -0,0 +1,27 @@
+ */
+
+/*
+ * This code demonstrates the use of threads and requires object
```
This comment seems out of date now

laurenmurphyx64 replied:
Good catch, changed

teburd commented on tests/subsys/llext/simple/src/object_ext.c:
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright (c) 2023 Intel Corporation.
```
2024 has arrived, I just caught myself today doing this :calendar:

laurenmurphyx64 replied:
Updated :clock3:

teburd reviewed (dismissed):
Minor nits

marc-hb commented:
This commit https://github.com/zephyrproject-rtos/zephyr/commit/ce243944375e46e94346bfb3dcbe8c499fc4996b ("llext: add object test case") and commit https://github.com/zephyrproject-rtos/zephyr/commit/1408d1e5b8cefcb0f4d52c1ffb41849f70826f39 ("tests: llext: compile architectures not supported yet") conflict with each other. They were submitted concurrently, tested separately and merged around the same time. All PRs fail now :-(

Please review hotfix #70406 urgently.
